src/guesslang_to_db.py

```diff
@@ -0,0 +1,178 @@
+"""
+Language Detector Script for Cloned Repositories
+
+This script provides functionality to analyse files within cloned
+repositories, detect their programming languages using
+the guesslang library, and write the results to a Supabase database.
+
+Run the script with the appropriate `--clone-dir` argument specifying the
+directory where repositories are cloned or let it use the default directory.
+
+Example:
+    $ python script.py --clone-dir /path/to/cloned_repos
+
+"""
+
+import argparse
+import os
+from pathlib import Path
+
+from loguru import logger
+from tqdm import tqdm
+from guesslang import Guess
+
+from db import read_from_db, write_to_db
+from utils.git_utils import get_working_directory_or_git_root
+
+# Configure logger
+logger.add("language_detector_script_log", rotation="500 MB")
+
+
+def parse_args():
+    """
+    This function defines the command-line arguments for specifying the directory
+    where repositories are cloned. The default directory is set to a subdirectory
+    named 'cloned_repo' within the project's 'data' folder.
+
+    Returns:
+    argparse.Namespace: Parsed command-line arguments.
+    """
+    parser = argparse.ArgumentParser(
+        description="Specify the directory for " "cloned repositories"
+    )
+    parser.add_argument(
+        "--clone-dir",
+        type=str,
+        default=str(
+            Path(get_working_directory_or_git_root()) / "data" / "cloned_repos"
+        ),
+        help="Directory where repositories are cloned. Defaults to "
+        "'data/cloned_repo' within the project's root directory.",
+    )
+    return parser.parse_args()
+
+
+def detect_language(file_path):
+    """
+    Detects the programming language of a given file using guesslang.
+    Args:
+        file_path (str): The path to the file.
+    Returns:
+        str: The name of the detected programming language.
+    """
+    logger.debug(f"Analysing file: {file_path}")
+
+    # Check if file exists
+    if not os.path.exists(file_path):
+        logger.error(f"File '{file_path}' not found.")
+        return "Unknown"
+
```

Comment:
I've run the code and realised that it's currently processing lots of files and folders unnecessarily. These records increase the noise in terms of a signal-to-noise ratio e.g. as the `.git` folder is currently being processed but shouldn't be as it's not part of the source code. Similarly `.zip` files are being processed (but fail as `guesslang` cannot parse their contents). 

Let's implement something similar to the excluded files list we have in the `list_test_files` function (`if item.suffix not in excluded_extensions:`) extended to also skip folders that begin with `.` (particularly `.git`) on the basis that many folders that start with a dot character aren't expected to have source files that belong to the codebase. (There are counter examples in a few repos). 

`.gitignore` files in a repo might also be worth processing in our code to generate the set of files to exclude from analysis by `guesslang`. Let's defer this work for now unless/until we can find a way to process these elegantly.

Reply:
- looked at the dataframe which I downloaded from the Supabase
- There are 42913 rows where the value for the `guesed_language` in unknown --> %11.68
- I excluded some extensions `['.png', '.gif', '.jpg', '.ttl', '.DS_Store', '.pcap', '.csv']` and also the file paths which included `/.git/` and excluded these from the original df
-  there's no file that ends with '.git' --> final_df[final_df['file_path'].str.endswith('.git')].shape[0]
- it reduced the unknown percentage to %3